src/pnpq/devices/odl_ozoptics_650ml.py

```diff
@@ -164,11 +165,19 @@ def serial_read(self, retries=10):
                 got_OK = True
             time.sleep(0.05)
             retries -= 1
+
+        if not got_OK:
+            # means command did not complete before soft timeout!
```

Comment:
What is a "soft timeout?"

Also, the logic on line 161 is problematic.

```
while self.conn.in_waiting > 0 or (got_OK is False and retries > 0):
```

Let's think about this.

If `in_waiting > 0`, then we can go through all 100 retries, and we can continue to retry 101, because `in_waiting > 0`. The first half of the `or` condition is still true.

As long as there is something `in_waiting`, the loop will continue.

Even if `got_OK == True`, as long as there is something `in_waiting`, the loop will continue.

This doesn't sound like the desired behavior.

I think that the `in_waiting` test should be removed from the if-condition.

Also, [don't use `is` to compare booleans, use `==`](https://stackoverflow.com/questions/4591125/is-it-safe-to-replace-with-is-to-compare-boolean-values). `is` is dangerous. Actually, here, you don't need either one.

```
while (not got_OK) and retries > 0:
```

Line 162 is also worrisome. What happens if `in_waiting` is 0?